src/k8s/pkg/k8sd/api/certs_refresh.go

```diff
@@ -0,0 +1,263 @@
+			commonName:   fmt.Sprintf("system:node:%s", snap.Hostname()),
+			organization: []string{"system:nodes"},
+			usages:       []v1.KeyUsage{v1.UsageDigitalSignature, v1.UsageKeyEncipherment, v1.UsageClientAuth},
+			signerName:   "k8sd.io/kubelet-client",
+			certificate:  &certificates.KubeletClientCert,
+			key:          &certificates.KubeletClientKey,
+		},
+		{
+			name:        fmt.Sprintf("k8sd-%d-worker-kube-proxy-client", req.Seed),
+			commonName:  "system:kube-proxy",
+			usages:      []v1.KeyUsage{v1.UsageDigitalSignature, v1.UsageKeyEncipherment, v1.UsageClientAuth},
+			signerName:  "k8sd.io/kube-proxy-client",
+			certificate: &certificates.KubeProxyClientCert,
+			key:         &certificates.KubeProxyClientKey,
+		},
+	} {
+		csr := csr
+		g.Go(func() error {
+			csrPEM, keyPEM, err := pkiutil.GenerateCSR(
+				pkix.Name{
+					CommonName:   csr.commonName,
+					Organization: csr.organization,
+				},
+				2048,
+				nil,
+				csr.hostnames,
+				csr.ips,
+			)
+			if err != nil {
+				return fmt.Errorf("failed to generate CSR for %s: %w", csr.name, err)
+			}
+
+			if _, err = client.CertificatesV1().CertificateSigningRequests().Create(ctx, &certificatesv1.CertificateSigningRequest{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: csr.name,
+				},
+				Spec: certificatesv1.CertificateSigningRequestSpec{
+					Request:    []byte(csrPEM),
+					Usages:     csr.usages,
+					SignerName: csr.signerName,
+				},
+			}, metav1.CreateOptions{}); err != nil {
+				return fmt.Errorf("failed to create CSR for %s: %w", csr.name, err)
+			}
+			watcher, err := client.CertificatesV1().CertificateSigningRequests().Watch(ctx, metav1.SingleObject(metav1.ObjectMeta{Name: csr.name}))
+			if err != nil {
+				return fmt.Errorf("failed to watch CSR %s: %w", csr.name, err)
+			}
+
+			defer watcher.Stop()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return nil
+				case evt, ok := <-watcher.ResultChan():
+					if !ok {
+						return fmt.Errorf("watch closed for CSR %s", csr.name)
+					}
+
+					request, ok := evt.Object.(*v1.CertificateSigningRequest)
+					if !ok {
+						return fmt.Errorf("expected a CertificateSigningRequest but received %#v", evt.Object)
+					}
+
+					approved, err := isCertificateSigningRequestApproved(request)
+					if err != nil {
+						return fmt.Errorf("csr is in an invalid state: %w", err)
+					}
+
+					if approved && isCertificateSigningRequestIssued(request) {
+						if _, _, err = pkiutil.LoadCertificate(string(request.Status.Certificate), ""); err != nil {
+							log.WithValues("csr", csr.name).Error(err, "CertificateSigningRequest failed")
+							return fmt.Errorf("CertificateSigningRequest failed: %w", err)
+						}
+						*csr.certificate = string(request.Status.Certificate)
+						*csr.key = string(keyPEM)
+						return nil
+					}
+				}
+			}
+		})
+
+	}
+
+	if err := g.Wait(); err != nil {
+		return response.InternalError(fmt.Errorf("failed to generate worker CSRs: %w", err))
```

Comment:
The `g` group does not generate CSRs, it waits for certificates to be approved and signed. a better message would probably `"failed to get worker node certificates: %w"`. Then the wrapped error will contain the cause (e.g. failed to generate, or CSR was denied/failed).

To be more precise, consider what some of the errors would look like:
- failed to generate worker CSRs: failed to generate CSR for $name <-- this is OK
- failed to generate worker CSRs: watch closed for CSR <-- not as OK, how is the watch failure related to generating the CSR?

Reply:
Looks much better, let me update the message.